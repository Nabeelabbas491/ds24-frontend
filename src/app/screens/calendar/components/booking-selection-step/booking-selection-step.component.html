<div class="flex flex-row justify-start bg-primary">
  <div class="flex flex-col justify-start items-center w-[56%] bg-neutral-n50">
    <ds-booking-user-profile [bookingTemplateSummary]="bookingTemplateSummary$ | async"></ds-booking-user-profile>
    <h4 class="font-medium my-3">{{ 'CALENDAR.SELECT_A_DATE' | translate }}</h4>
    <ds-calendar></ds-calendar>
  </div>
  <div class="w-[44%] bg-neutral-n0 p-5" *ngIf="selectedDay$ | async">
    <ds-booking-type-selection
      [meetingType]="meetingType$ | async"
      (meetingTypeChange)="meetingTypeChange($event)"
    ></ds-booking-type-selection>
    <div class="mt-8">
      <ds-time-selection></ds-time-selection>
    </div>
    <div class="mt-5" *ngIf="isBookingSelectionStepComplete$ | async">
      <button
        class="w-full h-10 py-2 px-4 bg-primary-p300-main text-neutral-n0 shadow-shadow-z200 text-center"
        (click)="nextStep()"
      >
        {{ 'BOOKING.BOOKING_OTHER.NEXT' | translate }}
      </button>
    </div>
  </div>
</div>
