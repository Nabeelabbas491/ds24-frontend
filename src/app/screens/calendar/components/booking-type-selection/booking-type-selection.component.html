<div class="mt-4">
  <h4 class="font-medium">{{ 'BOOKING.SELECT_BOOKING_TYPE' | translate }}</h4>
  <div class="mt-5" [formGroup]="form" *ngIf="meetingTypes$ | async as meetingTypes">
    <div
      class="h-14 bg-neutral-n0 shadow-shadow-z100 flex flex-col justify-center mb-3 cursor-pointer"
      *ngFor="let meetingType of meetingTypes"
    >
      <ds-radio [value]="meetingType.id" name="meetingtypradio" class="mx-3" formControlName="meetingType">
        <ds-form-label>{{ 'BOOKING.CALL_TYPE.' + meetingType.id | translate }}</ds-form-label>
      </ds-radio>
    </div>
  </div>
</div>
