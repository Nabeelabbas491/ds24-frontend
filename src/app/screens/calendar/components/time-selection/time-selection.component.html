<h4 class="font-medium">{{ 'BOOKING.SELECT_TIME' | translate }}</h4>
<div class="mt-5" *ngIf="selectedDay$ | async as selectedDay">
  <form [formGroup]="form">
    <ds-select formControlName="timeZone" class="text-left">
      <ds-select-filter></ds-select-filter>
      <ds-form-label>{{ 'BOOKING.SELECT_TIME_ZONE' | translate }}</ds-form-label>
      <ds-select-option *ngFor="let timezone of timeZones$ | async" [value]="timezone.name">{{
        'TIMEZONES.' + timezone.name | translate
      }}</ds-select-option>
    </ds-select>
  </form>
  <h4 class="text-neutral-n500 mt-3 text-base">
    {{ selectedDay?.dayMonthYearTitle | dateFormat: 'full_without_year' }}
  </h4>
  <ng-template #timeslotslist>
    <div class="flex flex-row justify-start flex-wrap gap-3 mt-3" *ngIf="bookingTimeSlots$ | async as timeSlots">
      <p *ngIf="!timeSlots || timeSlots.length === 0; else timeslots" class="text-neutral-n300">
        {{ 'BOOKING.BOOKING_OTHER.NOT_AVAILABLE_AT_THIS_DATE' | translate }}
      </p>
      <ng-template #timeslots>
        <ds-chip
          [selectable]="true"
          [size]="'default'"
          [chipColor]="chipColor.LightPrimaryBlue"
          [selected]="(selectedTimeSlot$ | async) === item"
          (chipChange)="selectTimeSlot(item)"
          *ngFor="let item of timeSlots"
        >
          <ds-chip-text class="w-20 text-center">{{ item.startTime }}</ds-chip-text>
        </ds-chip>
      </ng-template>
    </div>
  </ng-template>
  <ds-spinner
    [color]="spinnerColor"
    class="mx-auto pl-6"
    *ngIf="isBookingTimeSlotsPending$ | async; else timeslotslist"
    [size]="50"
  ></ds-spinner>
</div>
