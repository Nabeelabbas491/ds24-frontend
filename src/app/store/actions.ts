import { actions as calendar } from './slices/calendar.store';
import { actions as booking } from './slices/booking.store';
import { actions as bookingAPI } from './slices/booking-api.store';
import { actions as vendor } from './slices/vendor.store';
import { actions as notification } from './slices/notification.store';
import { actions as viewBreakPoint } from './slices/view-breakpoint.store';
import { actions as vendorBookingTemplate } from './slices/vendor-booking-template.store';
import { actions as vendorSettings } from './slices/vendor-settings.store';
import { actions as vendorAppointment } from './slices/vendor-appointment.store';
import { actions as vendorCalendar } from './slices/vendor-calendar.store';
import { actions as vendorSchedule } from './slices/vendor-schedule.store';
import { actions as timeZoneAPI } from './slices/timezone-api.store';
import { actions as googleIntegration } from './slices/google-integration.store';

export {
  calendar,
  booking,
  bookingAPI,
  vendor,
  notification,
  viewBreakPoint,
  vendorBookingTemplate,
  vendorSettings,
  vendorAppointment,
  vendorCalendar,
  vendorSchedule,
  timeZoneAPI,
  googleIntegration,
};
