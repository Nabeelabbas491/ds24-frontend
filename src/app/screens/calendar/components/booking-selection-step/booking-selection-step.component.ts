import { Component } from '@angular/core';
import { Store } from '@ngrx/store';
import { actions, selectors } from '../../../../store';
import { MeetingType } from 'src/app/types/misc.type';

@Component({
  selector: 'ds-booking-selection-step',
  templateUrl: './booking-selection-step.component.html',
})
export class BookingSelectionStepComponent {
  selectedDay$ = this.store.select(selectors.booking.selectSelectedDay);
  isBookingSelectionStepComplete$ = this.store.select(selectors.booking.isBookingSelectionStepComplete);
  meetingType$ = this.store.select(selectors.booking.selectMeetingType);
  bookingTemplateSummary$ = this.store.select(selectors.booking.selectBookingTemplateSummary);

  constructor(private store: Store) {}
  meetingTypeChange(meetingType: MeetingType) {
    this.store.dispatch(actions.booking.meetingType({ meetingType }));
  }

  nextStep() {
    this.store.dispatch(actions.booking.selectStep({ step: 'booking_form' }));
  }
}
