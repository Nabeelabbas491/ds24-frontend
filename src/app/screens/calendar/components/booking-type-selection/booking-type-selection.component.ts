import { Component, EventEmitter, Input, Output } from '@angular/core';
import { FormBuilder, FormControl, FormGroup } from '@angular/forms';
import { ReactiveLifecycles } from '@ds24/utilities';
import { map, switchMap, takeUntil } from 'rxjs';
import { Store } from '@ngrx/store';
import { selectors } from './../../../../store';
import { MeetingType } from './../../../../types/misc.type';

@Component({
  selector: 'ds-booking-type-selection',
  templateUrl: './booking-type-selection.component.html',
})
export class BookingTypeSelectionComponent extends ReactiveLifecycles {
  form!: FormGroup;

  @Input()
  meetingType: MeetingType | null = null;

  @Output()
  meetingTypeChange: EventEmitter<MeetingType> = new EventEmitter<MeetingType>();

  meetingTypes$ = this.store.select(selectors.bookingAPI.selectBookingMeetingTypes);

  constructor(
    private fb: FormBuilder,
    private store: Store,
  ) {
    super();
  }
  ngOnInit() {
    super.ngOnInit();

    this.form = this.fb.group({
      meetingType: new FormControl(this.meetingType?.id),
    });

    this.formControls.meetingType.valueChanges
      .pipe(
        switchMap(meetingTypeId =>
          this.meetingTypes$.pipe(map(meetingTypes => meetingTypes?.find(mt => mt.id === meetingTypeId))),
        ),
        takeUntil(this.ngOnDestroy$),
      )
      .subscribe((meetingType: MeetingType) => {
        this.meetingTypeChange.emit(meetingType);
      });
  }

  get formControls(): any {
    return this.form.controls;
  }
}
