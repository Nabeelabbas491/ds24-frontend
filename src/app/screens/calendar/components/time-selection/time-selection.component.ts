import { Component } from '@angular/core';
import { ChipColor, IconColor, IconName } from '@ds24/elements';
import { Store } from '@ngrx/store';
import { ReactiveLifecycles } from '@ds24/utilities';
import { FormBuilder, FormControl, FormGroup } from '@angular/forms';
import { takeUntil } from 'rxjs';

import * as actions from '../../../../store/actions';
import { selectors } from '../../../../store';
import { Slot } from './../../../../types/calendar-day.type';
@Component({
  selector: 'ds-time-selection',
  templateUrl: './time-selection.component.html',
})
export class TimeSelectionComponent extends ReactiveLifecycles {
  icons: typeof IconName = IconName;
  chipColor: typeof ChipColor = ChipColor;
  spinnerColor = IconColor.Primary300;

  selectedTimeZone$ = this.store.select(selectors.booking.selectTimeZone);
  selectedDay$ = this.store.select(selectors.booking.selectSelectedDay);
  selectedTimeSlot$ = this.store.select(selectors.booking.selectTimeSlot);

  isBookingTimeSlotsPending$ = this.store.select(selectors.bookingAPI.selectBookingTimeSlotsPending);
  bookingTimeSlots$ = this.store.select(selectors.bookingAPI.selectBookingTimeSlots);

  timeZones$ = this.store.select(selectors.timeZoneAPI.selectTimeZones);

  form: FormGroup = this.fb.group({
    timeZone: new FormControl(''),
  });

  constructor(
    private store: Store,
    private fb: FormBuilder,
  ) {
    super();
  }

  ngOnInit() {
    super.ngOnInit();

    this.selectedTimeZone$.pipe(takeUntil(this.ngOnDestroy$)).subscribe(x => {
      this.formControls.timeZone.setValue(x);
    });

    this.formControls.timeZone.valueChanges.pipe(takeUntil(this.ngOnDestroy$)).subscribe((timeZone: string) => {
      this.store.dispatch(actions.booking.timeZone({ timeZone }));
    });
  }
  selectTimeSlot(timeSlot: Slot) {
    this.store.dispatch(actions.booking.timeSlot({ timeSlot }));
  }

  get formControls(): any {
    return this.form.controls;
  }
}
