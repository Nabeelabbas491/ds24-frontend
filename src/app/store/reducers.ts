import { isDevMode } from '@angular/core';
import { MetaReducer, ActionReducerMap } from '@ngrx/store';
import { routerReducer } from '@ngrx/router-store';

import { State } from './state';

/**
 * Every reducer module's default export is the reducer function itself. In
 * addition, each module should export a type or interface that describes
 * the state of the reducer plus any selector functions. The `* as`
 * notation packages up all of the exports into a single object.
 */

import * as calendar from './slices/calendar.store';
import * as booking from './slices/booking.store';
import * as bookingAPI from './slices/booking-api.store';
import * as vendor from './slices/vendor.store';
import * as notification from './slices/notification.store';
import * as viewBreakPoint from './slices/view-breakpoint.store';
import * as vendorBookingTemplate from './slices/vendor-booking-template.store';
import * as vendorSettings from './slices/vendor-settings.store';
import * as vendorAppointment from './slices/vendor-appointment.store';
import * as vendorCalendar from './slices/vendor-calendar.store';
import * as vendorSchedule from './slices/vendor-schedule.store';
import * as timeZoneAPI from './slices/timezone-api.store';
import * as googleIntegration from './slices/google-integration.store';

/**
 * Our state is composed of a map of action reducer functions.
 * These reducer functions are called with each dispatched action
 * and the current or initial state and return a new immutable state.
 */
export const rootReducers: ActionReducerMap<State> = {
  [vendor.featureKey]: vendor.reducer,
  [calendar.featureKey]: calendar.reducer,
  [booking.featureKey]: booking.reducer,
  [bookingAPI.featureKey]: bookingAPI.reducer,
  [notification.featureKey]: notification.reducer,
  [viewBreakPoint.featureKey]: viewBreakPoint.reducer,
  [vendorBookingTemplate.featureKey]: vendorBookingTemplate.reducer,
  [vendorSettings.featureKey]: vendorSettings.reducer,
  [vendorAppointment.featureKey]: vendorAppointment.reducer,
  [vendorCalendar.featureKey]: vendorCalendar.reducer,
  [vendorSchedule.featureKey]: vendorSchedule.reducer,
  [timeZoneAPI.featureKey]: timeZoneAPI.reducer,
  [googleIntegration.featureKey]: googleIntegration.reducer,
  router: routerReducer,
};

/**
 * By default, @ngrx/store uses combineReducers with the reducer map to compose
 * the root meta-reducer. To add more meta-reducers, provide an array of meta-reducers
 * that will be composed to form the root meta-reducer.
 */
export const metaReducers: MetaReducer<State>[] = isDevMode() ? [] : [];
