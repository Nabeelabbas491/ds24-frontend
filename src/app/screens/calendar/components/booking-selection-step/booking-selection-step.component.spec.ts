import { ComponentFixture, TestBed } from '@angular/core/testing';

import { TranslateModule } from '@ngx-translate/core';
import { IconModule } from '@ds24/elements';
import { BookingTypeSelectionComponent } from '../booking-type-selection/booking-type-selection.component';
import { TimeSelectionComponent } from '../time-selection/time-selection.component';
import { FormsModule, ReactiveFormsModule } from '@angular/forms';
import { MockStore, provideMockStore } from '@ngrx/store/testing';
import { State, actions, selectors } from '../../../../store';
import { MemoizedSelector } from '@ngrx/store';
import { BookingSelectionStepComponent } from './booking-selection-step.component';
import { MeetingType } from './../../../../types/misc.type';

describe('BookingSelectionStepComponent', () => {
  let fixture: ComponentFixture<BookingSelectionStepComponent>;
  let instance: BookingSelectionStepComponent;
  let mockStore: MockStore<State>;

  let isBookingSelectionStepComplete: MemoizedSelector<State, boolean>;
  let meetingType: MemoizedSelector<State, MeetingType | null>;

  beforeEach(() => {
    TestBed.configureTestingModule({
      declarations: [BookingSelectionStepComponent, BookingTypeSelectionComponent, TimeSelectionComponent],
      imports: [TranslateModule.forRoot(), IconModule, FormsModule, ReactiveFormsModule],
      providers: [provideMockStore()],
    });
    fixture = TestBed.createComponent(BookingSelectionStepComponent);
    instance = fixture.componentInstance;
    mockStore = TestBed.inject(MockStore);

    isBookingSelectionStepComplete = mockStore.overrideSelector(
      selectors.booking.isBookingSelectionStepComplete,
      false,
    );
    meetingType = mockStore.overrideSelector(selectors.booking.selectMeetingType, {
      id: 2,
      name: 'Outbound Phone Call',
    });

    fixture.detectChanges();

    jest.spyOn(mockStore, 'dispatch');
  });

  it('should compile', () => {
    fixture.detectChanges();
    expect(fixture).toMatchSnapshot();
  });

  it('isBookingSelectionSepComplete should be set to true', () => {
    isBookingSelectionStepComplete.setResult(true);
    mockStore.refreshState();
    fixture.detectChanges();

    expect(fixture).toMatchSnapshot();
  });

  it('should demonstrate bookingType change ', () => {
    meetingType.setResult({ id: 1, name: 'Inbound Phone Call' });
    mockStore.refreshState();
    fixture.detectChanges();

    expect(fixture).toMatchSnapshot();
  });

  it('should dispatch next step action', () => {
    instance.nextStep();

    const action = actions.booking.selectStep({ step: 'booking_form' });

    expect(mockStore.dispatch).toHaveBeenLastCalledWith(action);
  });

  it('should create', () => {
    expect(instance).toBeTruthy();
  });
});
