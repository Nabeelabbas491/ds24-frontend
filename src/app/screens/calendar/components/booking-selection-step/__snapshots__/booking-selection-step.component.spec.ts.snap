// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BookingSelectionStepComponent isBookingSelectionSepComplete should be set to true 1`] = `
<ds-booking-selection-step
  bookingTemplateSummary$={[Function _Store]}
  isBookingSelectionStepComplete$={[Function _Store]}
  meetingType$={[Function _Store]}
  selectedDay$={[Function _Store]}
  store={[Function _MockStore]}
>
  <div
    class="flex flex-row justify-start bg-primary"
  >
    <div
      class="flex flex-col justify-start items-center w-[56%] bg-neutral-n50"
    >
      <ds-booking-user-profile />
      <h4
        class="font-medium my-3"
      >
        CALENDAR.SELECT_A_DATE
      </h4>
      <ds-calendar />
    </div>
  </div>
</ds-booking-selection-step>
`;

exports[`BookingSelectionStepComponent should compile 1`] = `
<ds-booking-selection-step
  bookingTemplateSummary$={[Function _Store]}
  isBookingSelectionStepComplete$={[Function _Store]}
  meetingType$={[Function _Store]}
  selectedDay$={[Function _Store]}
  store={[Function _MockStore]}
>
  <div
    class="flex flex-row justify-start bg-primary"
  >
    <div
      class="flex flex-col justify-start items-center w-[56%] bg-neutral-n50"
    >
      <ds-booking-user-profile />
      <h4
        class="font-medium my-3"
      >
        CALENDAR.SELECT_A_DATE
      </h4>
      <ds-calendar />
    </div>
  </div>
</ds-booking-selection-step>
`;

exports[`BookingSelectionStepComponent should demonstrate bookingType change  1`] = `
<ds-booking-selection-step
  bookingTemplateSummary$={[Function _Store]}
  isBookingSelectionStepComplete$={[Function _Store]}
  meetingType$={[Function _Store]}
  selectedDay$={[Function _Store]}
  store={[Function _MockStore]}
>
  <div
    class="flex flex-row justify-start bg-primary"
  >
    <div
      class="flex flex-col justify-start items-center w-[56%] bg-neutral-n50"
    >
      <ds-booking-user-profile />
      <h4
        class="font-medium my-3"
      >
        CALENDAR.SELECT_A_DATE
      </h4>
      <ds-calendar />
    </div>
  </div>
</ds-booking-selection-step>
`;
